File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseGameRole.java
package com.yahyeet.boardbook.model.firebase.repository;

import com.google.firebase.firestore.DocumentSnapshot;
import com.yahyeet.boardbook.model.entity.GameRole;
import com.yahyeet.boardbook.model.entity.GameTeam;

import java.util.HashMap;
import java.util.Map;

public class FirebaseGameRole extends AbstractFirebaseEntity<GameRole> {
    private String id;
    private String name;
    private String teamId;

    public FirebaseGameRole() {
    }

    public FirebaseGameRole(String id, String name, String teamId) {
        this.id = id;
        this.name = name;
        this.teamId = teamId;
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();

        if (name != null) {
            map.put("name", name);
        }

        if (teamId != null) {
            map.put("teamId", teamId);
        }

        return map;
    }

    @Override
    public GameRole toModelType() {
        GameRole gameRole = new GameRole(name);
        gameRole.setId(id);
        GameTeam gameTeam = new GameTeam();
        gameTeam.setId(teamId);
        gameRole.setTeam(gameTeam);
        return gameRole;
    }

    public static FirebaseGameRole fromModelType(GameRole gameRole) {
        return new FirebaseGameRole(gameRole.getId(), gameRole.getName(), gameRole.getTeam().getId());
    }

    public static FirebaseGameRole fromDocument(DocumentSnapshot document) {
        FirebaseGameRole firebaseGameRole = new FirebaseGameRole();
        firebaseGameRole.setId(document.getId());

        if (document.contains("name")) {
            firebaseGameRole.setName(document.getString("name"));
        }

        if (document.contains("teamId")) {
            firebaseGameRole.setTeamId(document.getString("teamId"));
        }

        return firebaseGameRole;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getTeamId() {
        return teamId;
    }

    public void setTeamId(String teamId) {
        this.teamId = teamId;
    }
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseGameRole.kt
package com.yahyeet.boardbook.model.firebase.repository

import com.google.firebase.firestore.DocumentSnapshot
import com.yahyeet.boardbook.model.entity.GameRole
import com.yahyeet.boardbook.model.entity.GameTeam

import java.util.HashMap

class FirebaseGameRole : AbstractFirebaseEntity<GameRole> {
    var name: String? = null
    var teamId: String? = null

    constructor() {}

    constructor(id: String, name: String, teamId: String) : super(id) {
        this.name = name
        this.teamId = teamId
    }

    override fun toMap(): Map<String, Any> {
        val map = HashMap<String, Any>()

        if (name != null) {
            map["name"] = name!!
        }

        if (teamId != null) {
            map["teamId"] = teamId!!
        }

        return map
    }

    override fun toModelType(): GameRole {
        val gameRole = GameRole(name)
        gameRole.id = id
        val gameTeam = GameTeam()
        gameTeam.id = teamId
        gameRole.team = gameTeam
        return gameRole
    }

    companion object {

        fun fromModelType(gameRole: GameRole): FirebaseGameRole {
            return FirebaseGameRole(gameRole.id, gameRole.name, gameRole.team!!.id)
        }

        fun fromDocument(document: DocumentSnapshot): FirebaseGameRole {
            val firebaseGameRole = FirebaseGameRole()
            firebaseGameRole.id = document.id

            if (document.contains("name")) {
                firebaseGameRole.name = document.getString("name")
            }

            if (document.contains("teamId")) {
                firebaseGameRole.teamId = document.getString("teamId")
            }

            return firebaseGameRole
        }
    }
}
